README.md
# wanji_laser_driver


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package wj_716_lidar
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
0.0.3 (2019-10-21)
-------------------
* 修改强度信息数据个数应与距离个数一致
* 修改距离无效值为无效值

0.0.2 (2019-10-14)
-------------------
* 修改角度配置範圍
* 增加默认port参数
* 增加最近距离限制

0.0.1 (2019-09-12)
-------------------
* 配置默认参数
* 增加发布base_link到laser tf
* add is_running service
* add heartbeat_pub_ msg




src/async_client.h
#ifndef ASYNC_CLIENT_H
#define ASYNC_CLIENT_H
#include <iostream>
#include <stdio.h>
#include <time.h>
#include <boost/asio.hpp>
#include <boost/bind.hpp>
#include <boost/thread.hpp>
#include <string>
#include <unistd.h>

using namespace std ;
using boost::asio::ip::udp;
using boost::asio::ip::tcp;
using namespace boost::asio;
#define MAX_LENGTH 50000
typedef void (*fundata_t)(const char* addr,int port,const char* data,const int len);
class Async_Client
{
public:
  Async_Client();
  Async_Client(boost::asio::io_service& io_service,ip::tcp::endpoint endpoint,fundata_t fundata_ );
  void client_async_write(char *buf,int len);
  void client_async_read();
  bool client_return_status();

private:
  void handle_read(const boost::system::error_code& error,
                   size_t bytes_transferred);

  void handle_write(const boost::system::error_code& error);
  io_service &iosev;
  ip::tcp::socket socket ;
  ip::tcp::endpoint ep ;
  boost::system::error_code ec;

  char        data_[MAX_LENGTH];
  bool        m_connected ;
  fundata_t   m_fundata ;
};

#endif // ASYNC_CLIENT_H


src/async_client.cpp
#include "async_client.h"

Async_Client::Async_Client(boost::asio::io_service& io_service,ip::tcp::endpoint endpoint,fundata_t fundata_ )
  :iosev(io_service),
    socket(iosev),
    ep(endpoint),
    m_connected(0),
    m_fundata(fundata_)
{
  socket.connect(ep,ec);
  if(ec)
  {
    std::cout << boost::system::system_error(ec).what() << std::endl;
    m_connected = 0 ;
  }
  else
  {
    cout<<" Connection Success! "<<ep.address().to_string()<<endl;
    m_connected = 1 ;
  }
  client_async_read();
}
void Async_Client::client_async_write(char *buf,int len)
{
  boost::asio::async_write(socket,
                           boost::asio::buffer(buf, len),
                           boost::bind(&Async_Client::handle_write, this,
                                       boost::asio::placeholders::error));
}
void Async_Client::client_async_read()
{
  socket.async_read_some(boost::asio::buffer(data_, MAX_LENGTH),
                         boost::bind(&Async_Client::handle_read, this,
                                     boost::asio::placeholders::error,
                                     boost::asio::placeholders::bytes_transferred));
}
bool Async_Client::client_return_status(){
  return m_connected ;
}


void Async_Client::handle_read(const boost::system::error_code& error,
                               size_t bytes_transferred)
{
  if (!error)
  {
    //printf("From %s Received data Len:%d\n",this->ep.address().to_string().c_str(),
          // bytes_transferred);
    (*m_fundata)(this->ep.address().to_string().c_str(),this->ep.port(),data_,bytes_transferred);
    this->client_async_read();
  }
  else
  {
    printf("%s\n",error.message().c_str());
    m_connected = 0 ;
  }
}

void Async_Client::handle_write(const boost::system::error_code& error)
{
  if (!error)
  {

  }
  else
  {
    printf("%s\n",error.message().c_str());
    m_connected = 0 ;
  }
}


src/wj_716_lidar_01.cpp
#include <ros/ros.h>
#include "async_client.h"
#include "wj_716_lidar_protocol.h"
using namespace wj_lidar;

int InitTcpConnection(const char* addr,int port,Async_Client **client_,fundata_t fundata_)
{

  io_service iosev;

  ip::tcp::endpoint ep(ip::address_v4::from_string(addr),port);
  *client_ = new Async_Client(iosev,ep,fundata_);

  iosev.run();

  return 1 ;
}

int boost_tcp_init_connection(const char* addr,int port,Async_Client **client_,fundata_t fundata_)
{
  int timecnt=0 ;
  *client_ = NULL ;
  boost::thread tmp(&InitTcpConnection,addr,port,client_,fundata_);
  tmp.detach() ;

  while(timecnt<50){
    timecnt++ ;
    usleep(20000); //20 ms
    if((*client_)->client_return_status()){
      return 0 ;
    }
  }
  *client_ = NULL ;
  return -1 ;

}

int boost_tcp_sync_send(Async_Client *client_ ,const char* msg,const int len)
{
  if(client_==NULL || client_->client_return_status()==0 ){
    printf("not connected , please connect first \n");
    return -1 ;
  }
  else{
    client_->client_async_write((char*)msg,len);
    return 0 ;
  }

  return 1;
}

int boost_tcp_sync_read(Async_Client *client_ )
{
  if(client_==NULL || client_->client_return_status()==0 ){
    printf("not connected , please connect first \n");
    return -1 ;
  }
  else{
    client_->client_async_read();
    return 0 ;
  }
  return 1;
}

/* ------------------------------------------------------------------------------------------
 *  show demo --
 * ------------------------------------------------------------------------------------------ */
wj_716_lidar_protocol *protocol;
Async_Client *client;
void CallBackRead(const char* addr,int port,const char* data,const int len)
{
  protocol->dataProcess(data,len);
}

void callback(wj_716_lidar::wj_716_lidarConfig &config,uint32_t level)
{
  protocol->setConfig(config,level);
}

//void timerCallback(const ros::TimerEvent&)
//{
//  char scandata[5]={0x02,0x02,0x02,0x02,0x00};
//  //cout << "20ms" <<endl;
//  boost_tcp_sync_send(client,scandata,5);
//}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "wj_716_lidar_01");
  ros::NodeHandle nh("~");
//  ros::Timer timer;
  std::string hostname;
  nh.getParam("hostname",hostname);
  cout << hostname <<endl;
  std::string port;
  nh.getParam("port",port);
  port = "2110";


  protocol = new wj_716_lidar_protocol();
  dynamic_reconfigure::Server<wj_716_lidar::wj_716_lidarConfig> server;
  dynamic_reconfigure::Server<wj_716_lidar::wj_716_lidarConfig>::CallbackType f;
  f = boost::bind(&callback,_1,_2);
  server.setCallback(f);
  client=NULL;

  boost_tcp_init_connection(hostname.c_str(),atoi(port.c_str()),&client,&CallBackRead);
//  timer= nh.createTimer(ros::Duration(0.02), timerCallback);   //ding shi fa song yao shu zhi ling
  boost_tcp_sync_read(client);

  ROS_INFO("Hello wj_716_lidar!");
  ros::spin();

}


src/wj_716_lidar_protocol.h
#ifndef WJ_716_LIDAR_PROTOCOL_H
#define WJ_716_LIDAR_PROTOCOL_H
#include <iostream>
#include "string.h"
#include <boost/shared_ptr.hpp>
#include <boost/asio.hpp>
#include <boost/asio/placeholders.hpp>
#include <boost/system/error_code.hpp>
#include <boost/bind/bind.hpp>
#include <ros/ros.h>
#include <visualization_msgs/Marker.h>
#include <sensor_msgs/LaserScan.h>
#include <dynamic_reconfigure/server.h>
#include <wj_716_lidar/wj_716_lidarConfig.h>
#include <tf/transform_broadcaster.h>
using namespace std ;
namespace wj_lidar
{
  #define MAX_LENGTH_DATA_PROCESS 20000
  typedef struct TagDataCache
  {
    char m_acdata[MAX_LENGTH_DATA_PROCESS];
    unsigned int m_u32in;
    unsigned int m_u32out;
  }DataCache;
  class wj_716_lidar_protocol
  {
  public:
    wj_716_lidar_protocol();
    bool dataProcess(const char *data,const int reclen);
    bool protocl(const char *data,const int len);
    bool OnRecvProcess(char *data, int len);
    bool checkXor(char *recvbuf, int recvlen);
    void send_scan(const char *data,const int len);
    ros::NodeHandle nh;
    ros::Publisher marker_pub;
    sensor_msgs::LaserScan scan;
    bool setConfig(wj_716_lidar::wj_716_lidarConfig &new_config,uint32_t level);
    //dynamic_reconfigure::Server<wj_safety_lidar_protocl::> dr_srv;
  private:
    char        data_[MAX_LENGTH_DATA_PROCESS];
    DataCache   m_sdata;
    wj_716_lidar::wj_716_lidarConfig config_;
    int m_n32PreFrameNo;
    float scandata[810];
    float scaninden[810];
		tf::TransformBroadcaster tf_broadcaster_;
  };
}
#endif // WJ_716_LIDAR_PROTOCOL_H


src/wj_716_lidar_protocol.cpp
#include "wj_716_lidar_protocol.h"
#include <iostream>

namespace wj_lidar
{
  bool wj_716_lidar_protocol::setConfig(wj_716_lidar::wj_716_lidarConfig &new_config,uint32_t level)
  {
    config_ = new_config;
    scan.header.frame_id = config_.frame_id;
    scan.angle_min = config_.min_ang;
    scan.angle_max = config_.max_ang;
    scan.angle_increment = config_.angle_increment;
    scan.time_increment = config_.time_increment;//0.00002469;
    scan.range_min = config_.range_min;
    scan.range_max = config_.range_max;
//    scan.ranges.resize(config_.resize);
    scan.intensities.resize(config_.resize);
    cout << "frame_id:" <<scan.header.frame_id<<endl;
    cout << "min_ang:" <<scan.angle_min<<endl;
    cout << "max_ang:" <<scan.angle_max<<endl;
    cout << "angle_increment:" <<scan.angle_increment<<endl;
    cout << "time_increment:" <<scan.time_increment<<endl;
    cout << "range_min:" <<scan.range_min<<endl;
    cout << "range_max:" <<scan.range_max<<endl;
    int resizeNum;
    cout << "resizeNum:" <<resizeNum<<endl;
    return true;
  }
   wj_716_lidar_protocol::wj_716_lidar_protocol()
   {
     memset(&m_sdata,0,sizeof(m_sdata));
     marker_pub = nh.advertise<sensor_msgs::LaserScan>("scan", 50);
     ros::Time scan_time = ros::Time::now();      //make a virtual data per sec
     scan.header.stamp = scan_time;

      m_n32PreFrameNo = 0;

      scan.header.frame_id = "laser";
      scan.angle_min = -2.35619449;
      scan.angle_max = 2.35619449;
      scan.angle_increment = 0.00582;
      scan.time_increment = 0.0667/1080;//0.00002469;
      scan.scan_time = 0.0666666701436;
      scan.range_min = 0;
      scan.range_max = 30;
      scan.ranges.resize(810);
      scan.intensities.resize(810);

      cout << "wj_716_lidar_protocl start success" << endl;

   }
   bool wj_716_lidar_protocol::dataProcess(const char *data, const int reclen)
   {
     if(reclen > MAX_LENGTH_DATA_PROCESS)
     {
       return false;
     }

     if(m_sdata.m_u32in + reclen > MAX_LENGTH_DATA_PROCESS)
     {
       cout<<"1.m_sdata.m_u32in + reclen > MAX_LENGTH_DATA_PROCESS"<<endl;
       memset(&m_sdata,0,sizeof(m_sdata));
       return false;
     }
     memcpy(&m_sdata.m_acdata[m_sdata.m_u32in],data,reclen*sizeof(char));
     m_sdata.m_u32in += reclen;
     while(m_sdata.m_u32out < m_sdata.m_u32in)
     {
       if(m_sdata.m_acdata[m_sdata.m_u32out] == 0x02 && m_sdata.m_acdata[m_sdata.m_u32out+1] == 0x02 &&
          m_sdata.m_acdata[m_sdata.m_u32out+2] == 0x02 && m_sdata.m_acdata[m_sdata.m_u32out+3] == 0x02)
       {
         unsigned l_u32reallen = (m_sdata.m_acdata[m_sdata.m_u32out + 4] << 24) |
                                 (m_sdata.m_acdata[m_sdata.m_u32out + 5] << 16) |
                                 (m_sdata.m_acdata[m_sdata.m_u32out + 6] << 8)  |
                                 (m_sdata.m_acdata[m_sdata.m_u32out + 7] << 0);
         l_u32reallen = l_u32reallen + 9;
         if(l_u32reallen <= (m_sdata.m_u32in - m_sdata.m_u32out + 1))
         {
           if(OnRecvProcess(&m_sdata.m_acdata[m_sdata.m_u32out],l_u32reallen))
           {
             m_sdata.m_u32out += l_u32reallen;
           }
           else
           {
             cout << "continuous frame"<<endl;
             int i;
             for(i == 1; i<l_u32reallen; i++)
             {
               if((m_sdata.m_acdata[m_sdata.m_u32out + i] == 0x02) &&
                  (m_sdata.m_acdata[m_sdata.m_u32out + i + 1] == 0x02) &&
                  (m_sdata.m_acdata[m_sdata.m_u32out + i + 2] == 0x02) &&
                  (m_sdata.m_acdata[m_sdata.m_u32out + i + 3] == 0x02))
               {
                 m_sdata.m_u32out += i;
                 break;
               }
               if(i == l_u32reallen)
               {
                 m_sdata.m_u32out += l_u32reallen;
               }
             }
           }
         }
         else if(l_u32reallen >= MAX_LENGTH_DATA_PROCESS)
         {
           cout << "l_u32reallen >= MAX_LENGTH_DATA_PROCESS"<<endl;
           m_sdata.m_u32out++;
           memset(&m_sdata,0,sizeof(m_sdata));
            m_n32PreFrameNo = 0;
         }
         else
         {
            m_n32PreFrameNo = 0;
           memset(&m_sdata,0,sizeof(m_sdata));
           cout << "unknow error"<<endl;
           break;
         }
       }
       else
       {
         m_sdata.m_u32out++;
       }
     } //end while(m_sdata.m_u32out < m_sdata.m_u32in)
     if(m_sdata.m_u32out >= m_sdata.m_u32in)
     {
       if(m_sdata.m_u32out > m_sdata.m_u32in)
         m_n32PreFrameNo = 0;
       if(m_sdata.m_u32in >= 10*896)
       {
             cout << "m_sdata.m_u32out = :"<<m_sdata.m_u32out<< "m_sdata.m_u32in = :"<<m_sdata.m_u32in<<endl;
             cout << "clear m_u32in ok"<<endl;
       }
       memset(&m_sdata,0,sizeof(m_sdata));

     }
      return true;
   }

   bool wj_716_lidar_protocol::OnRecvProcess( char *data, int len)
   {
     if(len > 0)
     {
       if(checkXor(data,len))
       {
         //cout << "entire data"<<endl;
         protocl(data,len);

       }
     }
     else
     {
       return false;
     }
     return true;
   }

   bool wj_716_lidar_protocol::protocl(const char *data, const int len)
   {
     if((data[8] == 0x73 && data[9] == 0x52)||(data[8] == 0x73 && data[9] == 0x53) )   //command type:0x73 0x52/0X53
     {
       static int s_n32ScanIndex,s_n32State;
       int l_n32PackageNo= (data[50] << 8) + data[51];                                        //shuju bao xu hao
       int l_n32FrameNo = (data[46]<<24) + (data[47]<<16) + (data[48]<<8) + data[49];        //quan hao

       if(l_n32PackageNo == 1)
       {
         s_n32ScanIndex = 0;
         m_n32PreFrameNo = l_n32FrameNo;

         for(int j = 0; j < 810;j=j+2)
         {
           scandata[s_n32ScanIndex] = (((unsigned char)data[85+j]) << 8) + ((unsigned char)data[86+j]);
           scandata[s_n32ScanIndex] /= 1000.0;
           if(scandata[s_n32ScanIndex]>=25)
           {
             scandata[s_n32ScanIndex]=0;
           }
           scan.intensities[s_n32ScanIndex] = 0;
           s_n32ScanIndex++;
         }
         s_n32State = 1;
       }
       else if(l_n32PackageNo == 2)
       {
         if(m_n32PreFrameNo == l_n32FrameNo)
         {
           for(int j = 0; j < 810;j=j+2)
           {
             scandata[s_n32ScanIndex] =(((unsigned char)data[85+j]) << 8) + ((unsigned char)data[86+j]);
             scandata[s_n32ScanIndex] /= 1000.0;
             if(scandata[s_n32ScanIndex]>=25)
             {
               scandata[s_n32ScanIndex]=0;
             }
            //  scan.intensities[s_n32ScanIndex] = 0;
             s_n32ScanIndex++;
           }

           // adjust angle_min to min_ang config param
           int index_min = (config_.min_ang+2.35619449)/scan.angle_increment;
           // adjust angle_max to max_ang config param
           int index_max =  810-((2.35619449-config_.max_ang)/ scan.angle_increment); //810 - 1;  //
           int laser_size = index_max-index_min;
           scan.ranges.resize(laser_size);
           scan.intensities.resize(laser_size);
            for (int j = index_min; j <= index_max; ++j)
            {
                if(scandata[j]>0.005)
                  scan.ranges[laser_size- 1-(j - index_min)] = scandata[j];
                else
                {
                  scan.ranges[laser_size- 1 - (j - index_min)] = std::numeric_limits<float>::infinity();
                }
                
                scan.intensities[laser_size- 1-(j - index_min)]=0;
            }
            //cout << "size:"<<scan.ranges.size()<< endl;

           ros::Time scan_time = ros::Time::now();      //make a virtual data per sec
           scan.header.stamp = scan_time;
           marker_pub.publish(scan);
         }
         else
         {
           s_n32ScanIndex = 0;
           s_n32State = 0;
           cout << "err:"<<l_n32FrameNo<< endl;
           cout << "m_n32PreFrameNo:"<<m_n32PreFrameNo<< endl;
           return false;
         }
       }

       return true;
     }
     else
     {
       return false;
     }

   }

   bool wj_716_lidar_protocol::checkXor( char *recvbuf,  int recvlen)
   {
     int i = 0;
     char check = 0;
     char *p = recvbuf;
     int len;
     if(*p == (char)0x02)
     {
       p = p+8;
       len = recvlen - 9;
       for(i = 0; i < len; i++)
       {
         check ^= *p++;
       }
       if(check == *p)
       {
         return true;
       }
       else
         return false;
     }
     else
     {
       return false;
     }
   }

}
